### ISSUE TITLE
Frontend Project Health Improvements

### ISSUE THREAD

**Issue by Bobgy** [kubeflow/pipelines#2407]
This is an umbrella issue to track action items to improve frontend code base project health.

Goal is to improve onboarding, developer experience and facilitate contribution.

Planning
* [x] p0 [Standardize code style using prettier](https://github.com/kubeflow/pipelines/issues/2406)
* [x] p0 [Tslint should ignore generated code](https://github.com/kubeflow/pipelines/issues/2404)
* [x] p0 [Upgrade typescript to 3.6](https://github.com/kubeflow/pipelines/pull/2428)
* [x] p0 0.2w [Refactor frontend to use promise based grpc-web metadata api](https://github.com/kubeflow/pipelines/issues/2196)
* [x] p0 0.1w [Document and fix recommended npm daily practices](https://github.com/kubeflow/pipelines/issues/2452)
* [x] ~~p0~~ p1 0.8w (because p0 issues already resolved) [Set up unit tests for frontend node server](https://github.com/kubeflow/pipelines/issues/2215)
* [x] p0 0.1w [Document how to set up frontend dev env using mock server/kubeflow deployment](https://github.com/kubeflow/pipelines/issues/2461)
* [x] p1 0.6w  [Migrate to create-react-app and typescript-eslint](https://github.com/kubeflow/pipelines/issues/2432)
* [x] p1 0.4w [Frontend node server strict type checking](https://github.com/kubeflow/pipelines/issues/2808)
* [ ] p2 0.4w Document how to run end-to-end tests locally https://github.com/kubeflow/pipelines/issues/3600
* [x] p1 [Typescript lint developer experience improvements](https://github.com/kubeflow/pipelines/issues/2429)
* [x] p1 [Upgrade node to 12 LTS](https://github.com/kubeflow/pipelines/issues/2386)
* [x] p1 [Enable optional chaining js feature](https://github.com/kubeflow/pipelines/pull/3154) (blocked by migrating to create-react-app)
* [x] p2 [[UI] Consider migrating from enzyme to react-testing-library](https://github.com/kubeflow/pipelines/issues/2732)
* [ ] p2 [Update frontend integration tests to work with latest selenium chrome](https://github.com/kubeflow/pipelines/issues/1462)
* [x] [Snapshot diff best practices](https://github.com/kubeflow/pipelines/pull/3166)

**Comment by Bobgy** [comment]
/area front-end
#frontend-pH

**Comment by Bobgy** [comment]
Hi @eterna2, do you want to add anything like pain points here?

**Comment by eterna2** [comment]
Looks great to me. I will add here if I think anything else.

Not sure if we wish to impose consistency in the usage of `async/await` vs `promise chaining` because I notice the node backend uses both.

**Comment by Bobgy** [comment]
@eterna2 Thanks for the suggestion. I've not written much code for node backend. Which do you prefer? I think `async/await` is clearer and easier to write by all means. We can recommend switching to it doing new changes.

Do you think we need a strict rule (maybe https://github.com/xjamundx/eslint-plugin-promise/blob/master/docs/rules/prefer-await-to-then.md)? We cannot use an eslint rule right now, but after we switch to eslint-typescript. Maybe we can use it.

**Comment by eterna2** [comment]
I generally prefer `async/await` over promises too because of the readability. 

Would nice to enforce this consistency just to reduce confusion on how we shld handle async functions.

> Just need to be careful when using them, as u can incorrectly blocks the thread w/o actually needing to.
>
> Here is a good article on it:
https://hackernoon.com/javascript-async-await-the-good-part-pitfalls-and-how-to-use-9b759ca21cda

**Comment by Bobgy** [comment]
> I generally prefer async/await over promises too because of the readability.

> Would nice to enforce this consistency just to reduce confusion on how we shld handle async functions.

I think that's a good idea. Do you want to contribute to this or any of the items above? I don't think I can have time for this refactoring shortly.

Rough ideas:
* We should first complete https://github.com/kubeflow/pipelines/issues/2460 for test coverage.
* Convert existing usage of promise chain to async/await.
* (optional) Add a code style document recommending best practices or enforce by lint rules (prefer automated approach). Maybe we don't even need this, when our code base only uses `async/await`, I think contributors will discover that consistency by themselves.

**Comment by jingzhang36** [comment]
I ran into an issue where FE complier complains about casing difference. We probably can remove the check at https://github.com/kubeflow/pipelines/blob/d315bf654c54d815d090ffff3ba0a47e60053c5e/frontend/tsconfig.json#L17

**Comment by Bobgy** [comment]
New issue that came out:
* [x] https://github.com/kubeflow/pipelines/pull/2672

**Comment by Bobgy** [comment]
Another new issue I found out:
* Inconsistency between building in `frontend/server` folder locally and in docker

**Comment by Bobgy** [comment]
Most important issues have all been fixed!!
/close
